File: develop/lab-3-dynamodb/dynamoDBPythonLab/reservations_statistics.py
# Copyright 2017 Amazon Web Services, Inc. or its affiliates. All rights
# reserved.

import boto3
from boto3.dynamodb.conditions import Key
import sys
from datetime import datetime, timedelta
import utils
import solution as dynamodb_solution

RESERVATIONS_TABLE_NAME = utils.LAB_S3_RESERVATIONS_TABLE_NAME
CITY_DATE_INDEX_NAME = "ReservationsByCityDate"


def query_by_city(city):
    print("City name is {0}".format(city))
    # Query Reservations table based on the input city and count the number of
    # reservations
    count_for_city = 0
    try:
        dynamodb = boto3.resource('dynamodb')
        recs = query_city_related_items(
            dynamodb, RESERVATIONS_TABLE_NAME, CITY_DATE_INDEX_NAME, city)

        # Retrieves and prints from recs dictionary returned by the query.
        for rec in recs['Items']:
            print("\t", rec['CustomerId'], rec['Date'])
        count_for_city = len(recs['Items'])
        print("Count of Reservations in the city: {0}".format(count_for_city))
    except Exception as err:
        print("Error Message: {0}".format(err))
    return count_for_city


def query_city_related_items(dynamodb, reservations_table_name, gsi_name, city):
    """Return the items returned by the query

    Keyword arguments:
    dynamodb -- DynamoDB resource
    reservations_table_name -- Table name
    gsi_name -- Table index name
    city -- City name
    """

    # TODO 2: Replace the solution with your own code
    return dynamodb_solution.query_city_related_items(
        dynamodb, reservations_table_name, gsi_name, city)

if __name__ == '__main__':
    print("Querying items by city")
    query_by_city(city="Reno")

File: develop/lab-3-dynamodb/dynamoDBJavaLab/src/main/java/com/amazonaws/lab/ReservationsStatistics.java
package com.amazonaws.lab;
// Copyright 2017 Amazon Web Services, Inc. or its affiliates. All rights reserved.

import java.util.Iterator;

import com.amazonaws.AmazonClientException;
import com.amazonaws.AmazonServiceException;
import com.amazonaws.services.dynamodbv2.AmazonDynamoDB;
import com.amazonaws.services.dynamodbv2.AmazonDynamoDBClientBuilder;
import com.amazonaws.services.dynamodbv2.document.DynamoDB;
import com.amazonaws.services.dynamodbv2.document.Item;
import com.amazonaws.services.dynamodbv2.document.ItemCollection;
import com.amazonaws.services.dynamodbv2.document.QueryOutcome;

// The ReservationsStatistics class queries the reservations table and reports the total number of reservations in a city.
public class ReservationsStatistics {

	// City for which number of reservations must be queried
	public static final String inputCity = "Reno";

	public static int itemCount = 0;

	private static final String RESERVATIONS_TABLE_NAME = ReservationsTableCreator.RESERVATIONS_TABLE_NAME;
	private static final String CITY_DATE_INDEX_NAME = ReservationsTableCreator.CITY_DATE_INDEX_NAME;

	private static DynamoDB dynamoDB = null;
	private static AmazonDynamoDB dynamoDBClient = null;

	public static void main(String[] args) throws Exception {

		try {
			// Instantiate DynamoDB client and object

			// Instantiate DynamoDB client and object
			dynamoDBClient = AmazonDynamoDBClientBuilder.standard().build();

			dynamoDB = new DynamoDB(dynamoDBClient);

			// Query items for a city
			ItemCollection<QueryOutcome> items = queryCityRelatedItems(dynamoDB, RESERVATIONS_TABLE_NAME,
					CITY_DATE_INDEX_NAME, inputCity);

			System.out.println("-------------------------------------------------------------");

			// Retrieve a reference to an iterator from the ItemCollection object returned
			// by the query.
			Iterator<Item> itemIterator = items.iterator();
			Item item = null;
			// Iterate over items
			while (itemIterator.hasNext()) {
				// Count the number of items
				itemCount++;
				item = itemIterator.next();
				System.out.printf("%s - %s - %s %n", item.getString("CustomerId"), item.getString("City"),
						item.getString("Date"));
			}
			System.out.println("-------------------------------------------------------------");
			System.out.printf("Summary: Number of reservations in %s city is %d %n", inputCity, itemCount);

		} catch (AmazonServiceException ase) {
			System.out.println("Error Message:    " + ase.getMessage());
			System.out.println("HTTP Status Code: " + ase.getStatusCode());
			System.out.println("AWS Error Code:   " + ase.getErrorCode());
			System.out.println("Error Type:       " + ase.getErrorType());
			System.out.println("Request ID:       " + ase.getRequestId());
		} catch (AmazonClientException ace) {
			System.out.println("Error Message: " + ace.getMessage());
		}
	}

	/**
	 * Query the reservations table to retrieve the items that relate to the given
	 * city
	 *
	 * @param dynamoDB
	 *            Instance of DynamoDB class
	 * @param reservationsTableName
	 *            Table name
	 * @param cityDateGsiName
	 *            Global secondary index
	 * @param inputCity
	 *            City
	 * @return Query results
	 */
	public static ItemCollection<QueryOutcome> queryCityRelatedItems(DynamoDB dynamoDB, String reservationsTableName,
			String cityDateGlobalSecondaryIndexName, String inputCity) {
		// TODO 2: Replace the solution with your own code
		return Solution.queryCityRelatedItems(dynamoDB, reservationsTableName, cityDateGlobalSecondaryIndexName, inputCity);
	}
}
